### PR TITLE
Allow promises to be constructed for non-default-constructible types

### PR DESCRIPTION
**Pull request by MitalAshok**
A fix for #2488 without breaking ABI compatibility.

For types that are not default constructible, `promise<T>` will now act like it was supposed to all along (`set_value` will copy/move construct a value rather than copy/move assign it). The value is stored like a std::optional, where `_Has_stored_result` is repurposed to be "has a value been constructed", and its old purpose is replaced with `_Has_stored_result || (_Exception != nullptr)`.

Differences are guarded with `is_default_constructible<_Ty>` to have the same behaviour as before, except that `promise<T>::set_exception(nullptr)` has been explicitly disallowed (It was already UB, since a precondition of `set_exception_ptr`/`set_exception_at_thread_exit` is "`p` is not null.", and would make `get` return a default constructed value).

I also removed a member function `void _Set_value(bool _At_thread_exit);` that had the comment `// store a (void) result` and its corresponding `_Set_value_raw` function, since they were unused and it would be a logical error to call them with a non-default-constructible type.

In vNext, `promise<T>` can be fixed for all types by replacing all the `is_default_constructible<_Ty>` checks with `false`.

### CODE REVIEW

File: stl/inc/future
Comment (frederick-vs-ja) [inline stl/inc/future:317]: I think MSVC STL tends to reduce tag dispatch by `if constexpr` (a C++17 feature usable as an extension in C++14 mode).
```suggestion
    template <typename _Ty2>
    void _Emplace_result(_Ty2&& _Val) {
        // TRANSITION: _Associated_state ctor will move or copy assign if _Result was default constructed
        // rather than move or copy construct
        if constexpr (is_default_constructible_v<_Ty>) {
            _Result = _STD forward<_Ty2>(_Val);
        } else {
            ::new (static_cast<void*>(_STD addressof(_Result._HeldValue))) _Ty(_STD forward<_Ty2>(_Val));
            _Has_stored_result = true;
        }
    }
```

File: stl/inc/future
Comment (frederick-vs-ja) [inline stl/inc/future:430]: May be better to use `if constexpr`.
```suggestion
        if constexpr (is_default_constructible_v<_Ty>) {
            _Has_stored_result = true;
        }
```

File: stl/inc/future
Comment (frederick-vs-ja) [inline stl/inc/future:380]: I think `is_default_constructible_v` is better in MSVC STL. Not sure whether `if constexpr` is clearer here.

File: stl/inc/future
Comment (StephanTLavavej) [reply stl/inc/future:317]: Yes, we strongly prefer `if constexpr` and use it in all new code. Its advantages are numerous (shorter and more readable source code, faster for the compiler front-end to instantiate, faster for the compiler back-end to deal with as it emits far less codegen). There's only one situation where `if constexpr` doesn't supersede tag dispatch and it's extremely rare (as `if constexpr` happens in a function body, it can't select between different member initializer lists in a constructor).

We're gradually cleaning up our codebase to convert old uses of tag dispatch to `if constexpr`, so if you see tag dispatch in existing code, that's why.

We have globally suppressed (within STL headers) the compiler warning that using `if constexpr` in C++14 is using Future Technology, so you can use it with no further issues.

File: stl/inc/future
Comment (StephanTLavavej) [reply stl/inc/future:430]: Yep, this would be a slight improvement for code clarity and debug codegen. (In non-optimized debug mode, a branch will be emitted at runtime for `if`, but not for `if constexpr`.)

File: stl/inc/future
Comment (StephanTLavavej) [reply stl/inc/future:380]: Yes, we strongly prefer using variable templates (shorter, clearer, and better compiler throughput).

I agree that using `if constexpr` here is a less clear-cut case. I would recommend doing so as it can avoid compiler warnings, and it makes the logic a bit clearer (separating the compile-time from the runtime conditions).

File: stl/inc/future
Comment (StephanTLavavej) [inline stl/inc/future:324]: I see that these are being removed because they were completely unused. That's fine (they're non-virtual and there are no downsides to deleting them), but I recommend editing your PR description to mention this removal.

File: stl/inc/future
Comment (StephanTLavavej) [inline stl/inc/future:356]: Thanks for adding this precondition check. The assertion message is accurate, but I think it would be helpful to additionally cite the Standardese, beyond mentioning what was called. Adding "is invalid, see N4901 [futures.promise]/20" would be sufficient. (Mostly I'm thinking of users with existing codebases who upgrade and see this assertion emitted - seeing an assertion without an explanation may make them think they've encountered a toolset bug, whereas a message citing the Standardese is more clearly diagnosing an existing problem in their code.)

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]
Thanks for fixing this long-standing bug! (I can't count how many times it's been reported over the years, but it's definitely "lots".) Main feedback here is a request for test coverage, plus minor codebase convention nitpicks. Your strategy looks good to me and it should preserve binary compatibility.

File: stl/inc/future
Comment (MitalAshok) [reply stl/inc/future]: Turns out we don't need a second member at all, and we can just have a union with no active member

**Comment by MitalAshok** [comment]
I've also (re)discovered this bug:

https://github.com/microsoft/STL/blob/afe0800cd6bf0d163bf678ad4e1feb3c130d1ec5/stl/inc/future#L265-L275

Also seen here: https://stackoverflow.com/a/42648634

Since `_Retrieved` is not set before the exception is thrown, the shared state isn't invalidated when `future::get` throws an exception, and the post condition [futures.unique.future]p17 "Postconditions: `valid() == false`" isn't held.

The fix would be a breaking change. I could have it fixed only with the new behaviour, or have the new behaviour follow the old behaviour in this edge case and just add a comment that it needs to be fixed.

**Comment by StephanTLavavej** [comment]
The `license/cla` check is stuck because there was a broken deployment of the CLA Bot which has been fixed. To get the check passing again, I am going to close and immediately reopen this PR.

File: tests/std/tests/GH_002488_promise_not_default_constructible_types/test.cpp
Comment (StephanTLavavej) [inline tests/std/tests/GH_002488_promise_not_default_constructible_types/test.cpp:68]: Naming suggestion: Although this isn't ambiguous/shadowing (because the test isn't `using namespace std;`), I recommend against choosing Standard identifiers for names in test code. Such avoidance increases readability by making Standard machinery be instantly recognizable (i.e. being able to assume that an `errc` refers to `std::errc`, especially because many of our tests are `using namespace std;`).

Here, `ec` or `code` would avoid the naming pseudo-collision.

File: tests/std/tests/GH_002488_promise_not_default_constructible_types/test.cpp
Comment (StephanTLavavej) [inline tests/std/tests/GH_002488_promise_not_default_constructible_types/test.cpp:82]: Similarly, I recommend choosing different names for these aliases. When I first glanced at this, I thought it was using C++17 Class Template Argument Deduction (where `std::future my_future = my_promise.get_future()` can deduce that you want a `std::future<int>`).

`Promise` and `Future` would be fine. `promise_t`, `future_t`, etc. would work too, or anything else.

File: tests/std/tests/GH_002488_promise_not_default_constructible_types/test.cpp
Comment (StephanTLavavej) [inline tests/std/tests/GH_002488_promise_not_default_constructible_types/test.cpp:158]: Suggestion: I believe that all of these lambdas written with `[]() -> T` could simply say `[]` as the `T` return type will be deduced, even in C++14 mode. However, if you like the explicit return type, feel free to keep it, it isn't harmful.

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]
Test looks good, thank you! One more set of small comments (the most significant being the Standard citation, which isn't very significant) and I think this will be ready to go :rocket:

**Review by StephanTLavavej — APPROVED** [review]

File: stl/inc/future
Comment (CaseyCarter) [inline stl/inc/future:215]: `is_default_constructible_v<const int>` is `true`, since `const int` can be value-initialized, but `default_initalizable<const int>` is `false` since `const int` cannot be default-initialized. Should we be using something like `default_initializable` (but obviously not a concept, since we need it in C++14 mode) to discriminate instead of `is_default_constructible_v`?

File: stl/inc/future
Comment (CaseyCarter) [inline stl/inc/future:193]: Am I missing something, or is this conversion operator unused? If not, can we remove it as extraneous?

File: stl/inc/future
Comment (CaseyCarter) [inline stl/inc/future:313]: Ditto present tense, and it's also confusing that this is talking about the `_Associated_state` constructor which has already completed. I _think_ we want:
```suggestion
        // TRANSITION, incorrectly assigns _Result when _Ty is default constructible
```

File: stl/inc/future
Comment (CaseyCarter) [inline stl/inc/future:292]: I think this could be a bit clearer (and who doesn't want to invalidate the future?)
```suggestion
        // TRANSITION: _Retrieved should be assigned before `_Exception` is thrown so that a `future::get`
        // that throws a stored exception invalidates the future (N4901 [futures.unique.future]/17)
```

File: tests/std/tests/GH_002488_promise_not_default_constructible_types/test.cpp
Comment (CaseyCarter) [inline tests/std/tests/GH_002488_promise_not_default_constructible_types/test.cpp:3]:
```suggestion
```

File: tests/std/tests/GH_002488_promise_not_default_constructible_types/test.cpp
Comment (CaseyCarter) [inline tests/std/tests/GH_002488_promise_not_default_constructible_types/test.cpp:12]:
```suggestion
```

File: tests/std/tests/GH_002488_promise_not_default_constructible_types/test.cpp
Comment (CaseyCarter) [inline tests/std/tests/GH_002488_promise_not_default_constructible_types/test.cpp:17]:
```suggestion
```

File: stl/inc/future
Comment (CaseyCarter) [reply stl/inc/future:215]: (If we do make a change here, we should add a test case for `promise<const int>` / `future<const int>`.)

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: stl/inc/future
Comment (CaseyCarter) [reply stl/inc/future:215]: After some discussion in Discord with @StephanTLavavej, we're happy to accept this as-is (without support for `const meow`) and we'll file a followup issue to make that improvement later. This is a _huge_ improvement over the status quo. I won't resolve this comment so we remember to file the followup, but feel free to ignore or implement it as you like.

Comment (StephanTLavavej) [reply stl/inc/future:215]: Filed #2599.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I've pushed a merge with `main` (no conflicts) followed by Casey's requested changes (I added backticks for consistency to one comment, no other changes). Should be ready to merge after @CaseyCarter's swift approval.

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks again for fixing this long-standing bug, and congratulations on your first microsoft/STL commit! :tada: :rocket: :heart_eyes_cat:

This will ship in VS 2022 17.2 Preview 3.

**Pull request merged**
